added support for Asrock IMB-1222 (Fintek F81804) and initial implementation for IMB-1213 (Nuvoton NCT6116D)

sronline opened the pull request:
started implementing support for Asrock IMB-1213 (F81804)

SamdCastle reviewed (changes requested):
I have an IMB-1215, will this update allow me to see my CPU fan? I can see temps, but can't see fan speed or RPM. Been trying to get Fan Control to work 😁🫶

Tellinq commented:
> I have an IMB-1215, will this update allow me to see my CPU fan? I can see temps, but can't see fan speed or RPM. Been trying to get Fan Control to work 😁🫶

Unlike the specific board that's being added, you get generic named sensors. However, if you have other software that you know of (like HWiNFO64) that shows a bit more detailed information or through some trial and error, I could help indicate properly named indexes.

To answer your question, you should be able to see the fan speeds/controls.

Tellinq commented on LibreHardwareMonitorLib/Hardware/Motherboard/SuperIOHardware.cs:
```diff
@@ -2183,6 +2185,64 @@ private static void GetFintekConfiguration(ISuperIO superIO, Manufacturer manufa
+                switch (model)
+                {
+                    case Model.IMB_1222:
+                        v.Add(new Voltage("+3V", 0, 150, 150));
+                        v.Add(new Voltage("+3VSB", 5, 150, 150));
+                        v.Add(new Voltage("VCore", 1));
+                        v.Add(new Voltage("VCCM", 2, 150, 150));
+                        v.Add(new Voltage("Voltage #4", 3, true));
+                        v.Add(new Voltage("Voltage #5", 4, true));
+                        v.Add(new Voltage("VBAT", 6, 150, 150));
+                        v.Add(new Voltage("Voltage #8", 7, true));
+                        v.Add(new Voltage("Voltage #9", 8, true));
+                        for (int i = 0; i < superIO.Temperatures.Length; i++)
+                        {
+                            if (i == 0)
+                            {
+                                t.Add(new Temperature("CPU Temperature", i));
+                            }
+                            else if (i == 1)
+                            {
+                                t.Add(new Temperature("M/B Temperature", i));
+                            }
+                            else
+                            {
+                                t.Add(new Temperature("Temperature #" + (i + 1), i));
+                            }
+                        }
+                        for (int i = 0; i < superIO.Fans.Length; i++)
+                            f.Add(new Fan("Fan #" + (i + 1), i));
+                        for (int i = 0; i < superIO.Controls.Length; i++)
+                            c.Add(new Control("Fan Control #" + (i + 1), i));
+                        break;
```
Some pointers I wanted to give out (I'm not a maintainer, but I do go out of my way to add motherboard-specific sensor support when possible):

# Naming schemes

You might get some requested changes on the naming scheme. Typically to avoid abbreviated names or duplicate namings.

If you want a good reference, I have a couple PRs for just that: #1364, #1360, and #1343

Some suggested naming changes:

`+3V` -> `+3.3V` (Typically the more accurate name)
`+3VSB` -> `+3V Standby`
`VBAT` -> `CMOS Battery`
`CPU Temperature` -> A more reflective name of the actual CPU portion that's being checked (`CPU Core`, `CPU Package`, or just generic `CPU` if unsure).
`M/B Temperature` -> `Motherboard`

# Temperatures
Some temperature sensors may be auxiliary sensors. From my experience with ASRock boards, it typically goes as the following:

Index 0: CPU (PECI) <- usually the proper index
Index 1: Either Motherboard or another potential CPU temperature, though commonly inaccurate
Index 2: Usually the Motherboard.

The rest of the sensors onward would be auxiliary sensors

Auxiliary sensors are typically "noise" sensors and are not worth showing.

Since you're adding a specific board. I recommend doing a bit of due diligence to check what sensors "officially" show up. Primarily in the UEFI and/or any software provided from ASRock that would show these temperatures.

Remember that even though the SuperIO chip may show up to a certain limit, the motherboard may not actually include that said amount (example the SuperIO chip can support 7 fans, but the motherboard may only use 3)

# Fans
Only thing I have to say is that I'd do a bit of a deeper look to properly label the type of fan (On which index is perhaps the CPU Fan, Chassis Fan, and Sata Power Fan (if I'm reading the motherboard labels correctly, it's a bit confusing with their label placement) set at?)

![image](https://github.com/user-attachments/assets/0316cfea-b6bc-48bc-bd7b-0c3e3a7102f1)

Also, if any of the fans aren't able to be controlled ("Power"/PSU fans and CPU Optional fans on some vendors may not have a control), I'd also adding those as a control if you ever encounter that.

sronline replied:
For voltages and temperatures I used the naming which is used in the BIOS:

![grafik](https://github.com/user-attachments/assets/761483d6-e0e2-48a9-9cb1-289534a60774)

SamdCastle commented:
> Unlike the specific board that's being added, you'll get generic named sensors. However, if you have other software that you know of (like HWiNFO64) that shows a bit more detailed information or through some trial and error, I could help indicate properly named indexes.
> 
> 
> 
> To answer your question, you should be able to see the fan speeds/controls.
> 
> 

I have tried about 5 programs but not had any luck with seeing anything about the CPU fan, yet I can see the fan speed in BIOS. Way out of my league so I appreciate your time!

SamdCastle commented:
Sorry to bother you again! I'm having trouble understanding how to apply these changes to Libre. I've managed to make the changes and build the new DLLs but when I replace them the program no longer opens.
Would you mind giving me a step by step guide of how to do it please? First time using GitHub, I'm almost certain its much simpler than i'm making it!

SamdCastle commented:
I'm just at work but will do so when I get home, thank you very much!

Yes that's probably best, I'm new to GitHub so wasn't sure how people do things, ideally I would've messaged you directly but couldn't see a way to do that, sorry!

Tellinq commented on LibreHardwareMonitorLib/Hardware/Motherboard/SuperIOHardware.cs:
```diff
@@ -2183,6 +2185,64 @@ private static void GetFintekConfiguration(ISuperIO superIO, Manufacturer manufa
+            case Manufacturer.ASRock:
+                switch (model)
+                {
+                    case Model.IMB_1222:
```
Wait, if you're saying in the PR title that this board is using NCT6116D, why is this under a Fintek configuration? Mistake?

Tellinq replied:
Gotcha. I did some work for another person that had an NCT6116D (IMB-1215). Might align with what the IMB-1213 uses but I don’t have that board so I can’t confirm https://github.com/Tellinq/LibreHardwareMonitor/commits/nct6116/

I can at least confirm adding the IMB-1215 data entry will work just fine as checked in https://github.com/LibreHardwareMonitor/LibreHardwareMonitor/discussions/1417 - check and see if the code I changed in my branch will work with the IMB-1213. Will need to add the IMB-1215 alongside the IMB-1215 to verify

sronline replied:
> Gotcha. I did some work for another person that had an NCT6116D (IMB-1215). Might align with what the IMB-1213 uses but I don’t have that board so I can’t confirm https://github.com/Tellinq/LibreHardwareMonitor/commits/nct6116/
> 
> I can at least confirm adding the IMB-1215 data entry will work just fine as checked in #1417 - check and see if the code I changed in my branch will work with the IMB-1213. Will need to add the IMB-1215 alongside the IMB-1215 to verify

I tried it and have one issue. Sometimes the temperatures of M/B and CPU are reported as 0

![grafik](https://github.com/user-attachments/assets/b3059f4a-ed26-47ee-b76f-81be5af67243)

t.Add(new Temperature("M/B Temperature", 1));
t.Add(new Temperature("CPU Temperature", 2));

sronline commented:
I implemented support for IMB-1213 and corrected the names of the sensors

![grafik](https://github.com/user-attachments/assets/25e92a43-e97f-4a12-bd7e-50b89522ded3)

![grafik](https://github.com/user-attachments/assets/f698ec22-f73d-422e-8c70-165d37103788)


with IMB-1213 there is one issue with the temperatures of system and cpu, sometime it is reported as 0. I don't know why this is the case

sronline commented:
I also added support for Asus Q87T:

https://github.com/LibreHardwareMonitor/LibreHardwareMonitor/commit/595aafacc0f018ce1972eff30afd7d4d93fd85c1

![grafik](https://github.com/user-attachments/assets/a9caaa65-e9be-4cd8-84ec-5b87d7c668a0)

PhyxionNL commented on .gitignore:
```diff
@@ -5,3 +5,4 @@
 .idea/
 .vs/
 packages/
+Backup/LibreHardwareMonitor.sln
```
Undo.

PhyxionNL commented on LibreHardwareMonitor.sln:
```diff
@@ -1,7 +1,7 @@
 Microsoft Visual Studio Solution File, Format Version 12.00
-# Visual Studio Version 16
-VisualStudioVersion = 16.0.29306.81
+# Visual Studio Version 17
+VisualStudioVersion = 17.11.35327.3
```
Undo.

PhyxionNL commented on LibreHardwareMonitorLib/LibreHardwareMonitorLib.csproj:
```diff
@@ -1,6 +1,6 @@
 <Project Sdk="Microsoft.NET.Sdk">
   <PropertyGroup>
-    <TargetFrameworks>net472;netstandard2.0;net6.0;net7.0</TargetFrameworks>
+    <TargetFrameworks>net472</TargetFrameworks>
```
Undo.

PhyxionNL commented on LibreHardwareMonitorLib/Hardware/Motherboard/SuperIOHardware.cs:
```diff
@@ -3661,6 +3782,37 @@ private static void GetNuvotonConfigurationD(ISuperIO superIO, Manufacturer manu
+                        v.Add(new Voltage("Voltage #14", 13, true));
+                        v.Add(new Voltage("Voltage #15", 14, true));
+                        t.Add(new Temperature("CPU Core", 0));
+                        t.Add(new Temperature("CPU (PECI Calibrated)", 1));
```
Just `PECI Calibrated` is OK here.

sronline commented:
I wasn't aware that I needed to resolve the proposed changes. Is there any chance of getting this merged now? Also, could the pull request be reopened?
